File: Dictionary.java
//-----------------------------------------------------------------------------
//Dictionary.java
//Binary Search Tree implementation of the Dictionary ADT
//-----------------------------------------------------------------------------

public class Dictionary implements DictionaryInterface {
   
   //Definition of Node
   private class Node {
      String key;
      String value;
      Node left;
      Node right;
      Node (String k, String v) {
         key = k;
         value = v;
         left = null;
         right = null;
      }
   }
   
   private int numPairs; //number of key/value pairs in the tree
   private Node root; //root of the tree
  

   // findKey()
   // returns the Node containing key k in the subtree rooted at R, or returns
   // null if no such Node exists
   private Node findKey(Node R, String k) {
      if (R == null || k.compareTo(R.key) == 0) {
         return R;
      }
      else if (k.compareTo(R.key) < 0) {
         return findKey(R.left, k);
      }
      else { //k.compareTo(R.key) > 0)
         return findKey(R.right, k);
      }
   }

   // findParent()
   // returns the parent of N in the subtree rooted at R, or returns null 
   // if N is equal to R. (pre: R != null)
   Node findParent(Node N, Node R) {
      Node P = null;
      if (N != R) {
         P = R;
         while ((P.left != N) && (P.right != N)) {
            if (N.key.compareTo(P.key) < 0) {
               P = P.left;
            }
            else
               P = P.right;
         }
      }
      return P;
   }
   
   // findLeftmost()
   // returns the leftmost Node in the subtree rooted at R, or null if R is null.
   Node findLeftmost (Node R) {
      if (R == null) {
         return R;
      }
      return findLeftmost(R.left);
   }
   
   // returns a string in order by location in the tree
   
   String stringInOrder(Node R) {
      String s = "";
      if (R != null) {
         s += stringInOrder(R.left);
         s += R.key+ " " +R.value+ "\n";
         s += stringInOrder(R.right);
      }
      return s;
   }
   
   public String toString () {
      return stringInOrder(root);
   }
   
   // deleteAll()
   // deletes all Nodes in the subtree rooted at N.
   void deleteAll (Node N) {
      if (N != null) {
         deleteAll(N.left);
         deleteAll(N.right);
      }
   } 
   
   // public functions -----------------------------------------------------------
   
   // Dictionary()
   // constructor for the Dictionary type

   public Dictionary() {
      root = null;
      numPairs = 0;
   }
   
   // isEmpty()
   // returns 1 (true) if D is empty, 0 (false) otherwise
   // pre: none
   public boolean isEmpty() {
      return (numPairs == 0);
   }
   
   // size()
   // returns the number of (key, value) pairs in D
   // pre: none
   public int size() {
      return(numPairs);
   }
   
   // lookup()
   // returns the value v such that (k, v) is in D, or returns NULL if no 
   // such value v exists.
   // pre: none
   public String lookup(String k) {
      Node N;
      N = findKey(root, k);
      return (N == null ? null : N.value);
   }
   
   // insert()
   // inserts new (key,value) pair into D
   // pre: lookup(k) = null
   public void insert(String k, String v) throws DuplicateKeyException {
      Node N, A, B;
      if (findKey(root, k) != null) {
         throw new DuplicateKeyException ("Dictionary Error: cannot insert() duplicate key: " + k);
      }
      N = new Node(k, v);
      B = null;
      A = root;
      while (A != null) {
         B = A;
         if (k.compareTo(A.key) < 0) {
            A = A.left;
         }
         else A = A.right;
      }
      if (B == null) {
         root = N;
      }
      else if (k.compareTo(B.key) < 0) {
         B.left = N;
      }
      else {
         B.right = N;
      }
      numPairs++;
   }
   
   // delete()
   // deletes pair with the key k
   // pre: lookup(D, k) != null
   public void delete (String k) throws KeyNotFoundException {
      Node N, P, S;
      N = findKey(root, k);
      if (N == null) {
         throw new KeyNotFoundException ("Dictionary Error: cannot delete() non-existent key: " + k);
      }
      if (N.left == null && N.right == null) {  // case 1 (no children)
         if (N == root){
            root = null;
         } else {
            P = findParent(N, root);
            if (P.right == N) {
               P.right = null;
            }
            else {
               P.left = null;
            }
         }
      } else if (N.right == null) {  // case 2 (left but no right child)
         if (N == root) {
            root = N.left;
         } else {
            P = findParent(N, root);
            if (P.right == N) {
               P.right = N.left;
            }
            else {
               P.left = N.left;
            }
         }
      } else if (N.left == null) {  // case 2 (right but no left child)
         if (N == root) {
            root = N.right;
         } else {
            P = findParent(N, root);
            if (P.right == N) {
               P.right = N.right;
            }
            else {
               P.left = N.right;
            }
         }
      } else {                     // case3: (two children: N->left!=NULL && N->right!=NULL)
         S = findLeftmost(N.right);
         N.key = S.key;
         N.value = S.value;
         P = findParent(S, N);
         if (P.right == S) {
            P.right = S.right;
         }
         else {
            P.left = S.right;
         }
      }
      numPairs--;
   }
   
   // makeEmpty()
   // re-sets D to the empty state.
   // pre: none
   public void makeEmpty() {
      deleteAll(root);
      root = null;
      numPairs = 0;
   }
}

File: Dictionary.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "Dictionary.h"

//structs and fields for NodeObj and DictionaryObj
typedef struct NodeObj {
   char* key;
   char* value;
   struct NodeObj* next;
} NodeObj;

typedef NodeObj* Node;

struct DictionaryObj {
   Node head;
   int numPairs;
} DictionaryObj;

void freeNode(Node* pN);

//Constructor for Dictionary
Dictionary newDictionary(void) {
   Dictionary D = malloc(sizeof(DictionaryObj));
   D->head = NULL;
   D->numPairs = 0;
   return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD) {
   if(*pD != NULL) {
      freeNode(&((*pD)->head));
      free(*pD);
      *pD = NULL;
   }
}

//Constructor for Nodes
Node newNode(char* key, char* value) {
   Node N = malloc(sizeof(NodeObj));
   N->key = key;
   N->value = value;
   return N;
}

//destructor for Nodes; frees Node to avoid memory leaks.
void freeNode(Node* pN) {
   if(*pN != NULL) {
       freeNode(&((*pN)->next));
       free(*pN);
       *pN = NULL;
   }
}

//Returns Node which contains key, else returns NULL.
Node findKey(Dictionary dict, char* key) {
   Node N = dict->head;
   for(int i = 0; i < dict->numPairs; i++) {
      if(strcmp(N->key, key) == 0) {
         return N;
      }
      N = N->next;
   }
   return NULL;
}

//Returns true if the dictionary contains no pairs; false otherwise.
int isEmpty(Dictionary dict) {
   return (size(dict) == 0);
}

int size(Dictionary dict) {
   return dict->numPairs;
}

// lookup()
// returns the value v such that (key, v) is in D, or returns NULL if no
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* key) {
   if(findKey(D, key) == NULL) {
      return NULL;
   }
   else {
      return findKey(D, key)->value;
   }
}

//Adds the pair (key,value) if there is no matching key; else creates an error.
void insert(Dictionary dict, char* key, char* value) {
   for(Node A = dict->head; A != NULL; A = A->next) {
      if(strcmp(A->key, key) == 0) {
         fprintf(stderr, "cannot insert duplicate keys");
         exit(EXIT_FAILURE);
      }
   }
   Node B = newNode(key, value);
   B->next = dict->head;
   dict->head = B;
   dict->numPairs++;
}

// delete()
// deletes pair with the key "key"
// pre: lookup(D, key)!=NULL
void delete(Dictionary D, char* key){
   Node previous = NULL;
   Node A;
   for(A = D->head; A != NULL; A = A->next) {
      if(strcmp(A->key, key) == 0) {
         if (A == D->head) {
            Node B = D->head;
            D->head = D->head->next;
            B->next = NULL;
            free(B);
         }
         else {
            previous->next = A->next;
            A->next = NULL;
            free(A);
         }
         break;
      }
      else if(A->next == NULL) {
         fprintf(stderr, "cannot delete non-existent key");
         exit(EXIT_FAILURE);
      }
      previous = A;
   }
   D->numPairs--;
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D) {
   freeNode(&(D->head));
   D->numPairs = 0;
}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D){
   Node N;
   for (N = D->head; N != NULL; N = N->next) {
      fprintf(out, "%s %s\n", N->key, N->value);
   }
}
